### DISCUSSION TITLE
About preventing automated failover from happening when running out of storage

### DISCUSSION THREAD

**Discussion by gbartolini** [cloudnative-pg/cloudnative-pg#3775] [Ideas]
At the moment, when PostgreSQL runs out of disk while writing WAL files, it shuts down. The readiness probe starts failing and CloudNativePG initiates a failover. Considering that all instances in a HA cluster have the same characteristics, the same problem will eventually rise in the promoted standby, generating an avalanche effect with no other solution than manual intervention.

While we try and solve this through automated scaling of volumes (we are actively working on this issue as members of DoK with other vendors, including Percona - see https://percona.community/blog/2024/01/10/data-on-kubernetes-community-initiatives/), there are already a couple of things we could do to improve the situation.

For example, we could introduce a default behavior that pre-emptively hibernates the cluster in case of storage exhaustion, and properly logs the information. Or, if we think this is too much, we can add an option (I honestly do not see how this can improve the overall situation, but I'd like to hear more comments).

Another option - which is actually a trick that many of us database people used to do - is to leave a 'buffer file' (for example, with a size equivalent to a few WAL files) in the volume where WAL files are located. And remove this file before shutting down the instance.

Other ideas?

**Comment by mnencia** [comment]
I think that hibernating the cluster is too surprising and unnecessary. We could just fence the primary and report the issue.

**Reply by gbartolini** [reply]
You convinced me. It actually make more sense, provided we are able to easily highlight this in the status of the cluster.

**Reply by MichaluxPL** [reply]
I totally agree with @mnencia. Fenicing would be enough to deal with such a situation. Hibernation seems an overkill :) and would slow down cluster revival (after dealing with storage exhaustion situation). After adding additional space it would require just to unfence the cluster to restore it to life.
Leaving a buffer file would be also a good idea too, but I would consider two things for such a solution:
a) count buffer file size according to WAL storage size or make it's size configurable by the user
b) make it optional (additional manifest parameter ?)

BTW: In my opinion similar behavior (fencing) should also take place in case of data storage exhaustion.
Of course in both caseses clear log info would be nice :) I'd also add a relevant info in cnpg plugin status.

**Reply by nareshov** [reply]
I like the idea of an emergency ballast file. Here's another popular DB adopting such a strategy: https://www.cockroachlabs.com/docs/stable/cluster-setup-troubleshooting#storage-issues

+1 to fencing over hibernation.

**Reply by gbartolini** [reply]
@nareshov thanks so much for this pointer!

**Comment by divyangjp** [comment]
How long before the auto expansion of storage becomes reality?

**Reply by gbartolini** [reply]
I can't say now. We will most likely meet at KubeCon in Paris and discuss.

**Comment by gbartolini** [comment]
I was able to trigger the issue of exhaustion of storage where WAL files are written.

I used the standard `cluster-example.yaml` file, with a single storage of 1GB:

```sh
kubectl apply -f https://cloudnative-pg.io/documentation/current/samples/cluster-example.yaml
```

Created a `BALLAST` file:

```sh
kubectl exec -ti cluster-example-1 -c postgres \
  -- dd if=/dev/zero of=/var/lib/postgresql/data/pgdata/pg_wal/BALLAST bs=1M count=16
```

Which returned:

```console
16+0 records in
16+0 records out
16777216 bytes (17 MB, 16 MiB) copied, 0.0109503 s, 1.5 GB/s
```

Then filled the `PGDATA` to run out of disk space, with:

```bash
kubectl exec -ti cluster-example-1 -c postgres \
  -- dd if=/dev/zero of=/var/lib/postgresql/data/pgdata/pg_wal/WASTE bs=1M count=1024
```

Which returned this:

```console
dd: error writing '/var/lib/postgresql/data/pgdata/pg_wal/WASTE': No space left on device
816+0 records in
815+0 records out
854978560 bytes (855 MB, 815 MiB) copied, 5.54574 s, 154 MB/s
```

I checked both space and `pg_controldata`'s state:

```console
$ kubectl exec -ti cluster-example-1 -c postgres -- pg_controldata | grep state
Database cluster state:               in production

$ kubectl exec -ti cluster-example-1 -c postgres -- df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay          80G  5.0G   76G   7% /
tmpfs            64M     0   64M   0% /dev
tmpfs           7.7G     0  7.7G   0% /sys/fs/cgroup
/dev/nvme0n1p1   80G  5.0G   76G   7% /run
tmpfs            15G  1.1M   15G   1% /dev/shm
tmpfs            15G   36K   15G   1% /etc/app-secret
/dev/nvme1n1    974M  958M     0 100% /var/lib/postgresql/data
tmpfs            15G   12K   15G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs           7.7G     0  7.7G   0% /proc/acpi
tmpfs           7.7G     0  7.7G   0% /sys/firmware
```

Then manually ran:

```sh
kubectl exec -ti cluster-example-1 -c postgres -- psql -c 'SELECT pg_switch_wal()'
```

This will trigger a failover. Be quick and get the log of `cluster-example-1` before the pod restarts, by running:

```sh
kubectl logs cluster-example-1 > cluster-example-1.json.log
```

In case the pod already restarted, get the previous log with:

```sh
kubectl logs -p cluster-example-1 > cluster-example-1.json.log
```

You should get something similar to the attached [cluster-example-1.json.log](https://github.com/cloudnative-pg/cloudnative-pg/files/14483689/cluster-example-1.json.log) file.

What is important to note is that unfortunately, only the log contains the information about the `PANIC`:

```json
{
  "level": "info",
  "ts": "2024-03-04T15:23:53Z",
  "logger": "postgres",
  "msg": "record",
  "logging_pod": "cluster-example-1",
  "record": {
    "log_time": "2024-03-04 15:23:53.168 UTC",
    "process_id": "25",
    "session_id": "65e5e64e.19",
    "session_line_num": "1",
    "session_start_time": "2024-03-04 15:18:38 UTC",
    "transaction_id": "0",
    "error_severity": "PANIC",
    "sql_state_code": "53100",
    "message": "could not write to file \"pg_wal/xlogtemp.25\": No space left on device",
    "backend_type": "walwriter",
    "query_id": "0"
  }
}
```

What I think we should do is watch for any `PANIC` message with `could not write to file \"pg_wal/xlogtemp.XXXXXX\": No space left on device`. If we detect that the primary has been shut down and the log contains this message, we should fence the cluster. Is this possible?

**Reply by NiharDudam** [reply]
Instead of reacting late and monitoring pglogs, can we monitor df or du output of the pgdata and pgwal directories and act on a certain threshold? The threshold can be configured at CR level ?

**Reply by gbartolini** [reply]
We still need to implement this from happening. I agree with you we then need to coordinate with autoscaling (see how this discussion originated). Although our operator is capable of automatically manage resizing, we should question whether it makes more sense for us to go down that path or integrate ourselves with Kubernetes level solutions?

Also, if we have WAL files in a separate volume, we already have everything we need to intervene at a certain threshold. I am talking about the `cnpg_collector_pg_wal` metric. In particular, we have:

- `count` representing the number of WAL files currently in the WAL volume
- `volume_max` representing the maximum number of WAL files that the WAL volume is able to store

The rest is math. This could be the first level of monitoring. The issue persist with PGDATA storing WAL files. In that case, we need to use more heavy disk checks.

**Comment by gbartolini** [comment]
Covered by #4521

**Reply by gbartolini** [reply]
MERGED!

**Comment by gbartolini** [comment]
Although this discussion goes beyond what's been merged in #4404, I am closing this discussion as the problem has been solved and the solution will be part of 1.24.

**Comment by pkutakovvk** [comment]
Hello Gabriel,

I've discovered this thread pretty late, anyway let me ask a question.
While fencing is great to prevent upcoming data loss and corruption, it makes database unavailable for the application. Yes, if you have a cluster and application is using <cluster-name>-ro  service to do reading, it will look better, but generally if you use only one connection to the master it will be unavailable.
What I would prefer as a database user - switch DB to readonly mode when disk space is exhausted.
This can be achieved via setting "default_transaction_read_only" system option (and stop all running write queries of course, maybe just forcibly drop all active connections).

What do you think? You definitely discussed that option internally...
